testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js

```diff
@@ -1819,6 +1866,295 @@ QUnit.module('Options', {
+                    selectionMode: 'range'
+                });
+                const expectedValues = [new Date('2023/01/13'), new Date('2023/01/15')];
+                const $cell = $(getCurrentViewInstance(this.calendar).$element().find('*[data-value="2023/01/15"]'));
+
+                $cell.trigger('dxclick');
+
+                assert.deepEqual(this.calendar.option('values'), expectedValues);
+            });
+
+            QUnit.test('Should swap startDate and endDate on cell when clicked endDate is less then startDate', function(assert) {
+                this.reinit({
+                    values: ['2023/01/13', null],
+                    selectionMode: 'range'
+                });
+                const expectedValues = [new Date('2023/01/07'), new Date('2023/01/13')];
+                const $cell = $(getCurrentViewInstance(this.calendar).$element().find('*[data-value="2023/01/07"]'));
+
+                $cell.trigger('dxclick');
+
+                assert.deepEqual(this.calendar.option('values'), expectedValues);
+            });
+
+            [
+                {
+                    values: [null, null],
+                    scenario: 'when both values are not defined'
+                },
+                {
+                    values: ['2023/01/13', '2023/01/17'],
+                    scenario: 'when both values are defined'
+                }
+            ].forEach(({ values, scenario }) => {
+                QUnit.test(`Cells should not have ${CALENDAR_RANGE_DATE_CLASS} class on hover ${scenario}`, function(assert) {
+                    if(devices.real().deviceType !== 'desktop') {
+                        assert.ok(true, 'test does not actual for mobile devices');
+                        return;
+                    }
+
+                    this.reinit({
+                        values,
+                        selectionMode: 'range'
+                    });
+                    const $cell = $(getCurrentViewInstance(this.calendar).$element().find('*[data-value="2023/01/25"]'));
+
+                    $cell.trigger('mouseenter');
+
+                    assert.notOk($cell.hasClass(CALENDAR_RANGE_DATE_CLASS));
+                });
+            });
+
+            QUnit.test(`Cells should have ${CALENDAR_RANGE_DATE_CLASS} class on hover when only startDate is defined`, function(assert) {
+                if(devices.real().deviceType !== 'desktop') {
+                    assert.ok(true, 'test does not actual for mobile devices');
+                    return;
+                }
+
+                this.reinit({
+                    values: ['2023/01/13', null],
+                    selectionMode: 'range'
+                });
+                const $cell = $(getCurrentViewInstance(this.calendar).$element().find('*[data-value="2023/01/15"]'));
+
+                $cell.trigger('mouseenter');
+
+                assert.ok($cell.hasClass(CALENDAR_RANGE_DATE_CLASS));
+            });
+        });
+    });
```

Comment:
shouldn't we add a couple of tests for a runtime change?

Reply:
what tests do u mean, there are tests for runtime switching from single to multi/range and vice verse